src/main/native/org_cryptomator_windows_keychain_WinHello_Native.cpp

```diff
@@ -0,0 +1,228 @@
+    auto dataBuffer = CryptographicBuffer::CreateFromByteArray(array_view<const uint8_t>(cleartextVec.data(), cleartextVec.size()));
+    auto encryptedBuffer = CryptographicEngine::Encrypt(aesKey, dataBuffer, keyMaterial);
+
+    return vectorToJbyteArray(env, iBufferToVector(encryptedBuffer));
+
+  } catch (winrt::hresult_error const& hre) {
+    HRESULT hr = hre.code();
+    winrt::hstring message = hre.message();
+    std::wcout << L"Error: " << message.c_str() << L" (HRESULT: 0x" << std::hex << hr << L")" << std::endl;
+    auto byteArray = env->NewByteArray(0);
+    return byteArray;
+  } catch (const std::exception& e) {
+    std::cout << "Warning: " << e.what() << std::endl;
+    auto byteArray = env->NewByteArray(0);
+    return byteArray;
+  } catch (...) {
+    std::cout << "Caught an unknown exception" << std::endl;
+    auto byteArray = env->NewByteArray(0);
+    return byteArray;
+  }
+}
+
+// Decrypts data using Windows Hello KeyCredentialManager API and derived key from signed challenge
+jbyteArray JNICALL Java_org_cryptomator_windows_keychain_WinHello_00024Native_getEncryptionKey
+(JNIEnv* env, jobject obj, jbyteArray ciphertext, jbyteArray challenge) {
+  queueSecurityPromptFocus();
+  try {
+    // Convert Java byte arrays to C++ vectors
+    std::vector<uint8_t> ciphertextVec = jbyteArrayToVector(env, ciphertext);
+    std::vector<uint8_t> challengeVec = jbyteArrayToVector(env, challenge);
+
+    // Take the random challenge and sign it by Windows Hello
+    // to create the key. The challenge is also used as the IV.
+    IBuffer key;
+    if (!deriveEncryptionKey(challengeVec, key)) {
+      throw std::runtime_error("Failed to generate the encryption key with the Windows Hello credential.");
+    }
+
+    auto algorithmName = SymmetricAlgorithmNames::AesCbc();
+    auto aesProvider = SymmetricKeyAlgorithmProvider::OpenAlgorithm(algorithmName);
+    auto keyMaterial = CryptographicBuffer::CreateFromByteArray(array_view<const uint8_t>(key.data(), key.Length()));
+    auto aesKey = aesProvider.CreateSymmetricKey(keyMaterial);
+    auto dataBuffer = CryptographicBuffer::CreateFromByteArray(array_view<const uint8_t>(ciphertextVec.data(), ciphertextVec.size()));
+    auto decryptedBuffer = CryptographicEngine::Decrypt(aesKey, dataBuffer, keyMaterial);
+
+    return vectorToJbyteArray(env, iBufferToVector(decryptedBuffer));
+
+  } catch (winrt::hresult_error const& hre) {
+    HRESULT hr = hre.code();
+    winrt::hstring message = hre.message();
+    std::wcout << L"Error: " << message.c_str() << L" (HRESULT: 0x" << std::hex << hr << L")" << std::endl;
+    auto byteArray = env->NewByteArray(0);
+    return byteArray;
+  } catch (const std::exception& e) {
+    std::cout << "Warning: " << e.what() << std::endl;
+    auto byteArray = env->NewByteArray(0);
+    return byteArray;
+  } catch (...) {
+    std::cout << "Caught an unknown exception" << std::endl;
+    auto byteArray = env->NewByteArray(0);
+    return byteArray;
+  }
+}
```

Comment:
Refactoring this would save 7 lines of duplicate code, but would make reading the code harder. As we are talking encryption here, the code is already hard enough to understand. I'll keep it as it is.